Add test scenarios for nullable return types in route handlers for RDG

Validate scenarios where return types in user's route handlers might be marked as nullable:

- `ValueTask<T>?`/`Task<T>?`
- `T?`

Found a bug:

```csharp
Todo? GetMaybeTodo(int id) => id == 0 ? null : new Todo(1, "foo");
app.MapGet("/", GetMaybeTodo);
```

Produces:

```
/Users/captainsafia/repos/aspnetcore/src/Http/samples/MinimalSample/Microsoft.AspNetCore.Http.Generators/Microsoft.AspNetCore.Http.Generators.RequestDelegateGenerator/GeneratedRouteBuilderExtensions.g.cs(107,100): error CS8639: The typeof operator cannot be used on a nullable reference type [/Users/captainsafia/repos/aspnetcore/src/Http/samples/MinimalSample/MinimalSample.csproj]
```

--- a/src/Http/Http.Extensions/gen/StaticRouteHandlerModel/StaticRouteHandlerModel.Emitter.cs
+++ b/src/Http/Http.Extensions/gen/StaticRouteHandlerModel/StaticRouteHandlerModel.Emitter.cs
@@ -84,21 +84,28 @@ internal static class StaticRouteHandlerModelEmitter
         {
             return;
         }
+        if (endpoint.Response.IsAwaitable)
+        {
+            codeWriter.WriteLine($"var task = handler({endpoint.EmitArgumentList()});");
+        }
+        if (endpoint.Response.IsAwaitable && endpoint.Response.ResponseType?.NullableAnnotation == NullableAnnotation.Annotated)
+        {
+            codeWriter.WriteLine("if (task == null)");
+            codeWriter.StartBlock();
+            codeWriter.WriteLine("""throw new InvalidOperationException("The Task returned by the Delegate must not be null.");""");
+            codeWriter.EndBlock();
+        }
         if (!endpoint.Response.HasNoResponse)
         {
             codeWriter.Write("var result = ");
         }
-        if (endpoint.Response.IsAwaitable)
-        {
-            codeWriter.Write("await ");
-        }
-        codeWriter.WriteLine($"handler({endpoint.EmitArgumentList()});");
+        codeWriter.WriteLine(endpoint.Response.IsAwaitable ? "await task;" : $"handler({endpoint.EmitArgumentList()});");
 
         endpoint.Response.EmitHttpResponseContentType(codeWriter);
 
         if (!endpoint.Response.HasNoResponse)
         {
-            codeWriter.WriteLine(endpoint.Response.EmitResponseWritingCall(endpoint.IsAwaitable));
+            endpoint.Response.EmitResponseWritingCall(codeWriter, endpoint.IsAwaitable);
         }
         else if (!endpoint.IsAwaitable)
         {
@@ -125,33 +132,37 @@ internal static class StaticRouteHandlerModelEmitter
         }
     }
 
-    private static string EmitResponseWritingCall(this EndpointResponse endpointResponse, bool isAwaitable)
+    private static void EmitResponseWritingCall(this EndpointResponse endpointResponse, CodeWriter codeWriter, bool isAwaitable)
     {
         var returnOrAwait = isAwaitable ? "await" : "return";
 
         if (endpointResponse.IsIResult)
         {
-            return $"{returnOrAwait} GeneratedRouteBuilderExtensionsCore.ExecuteAsyncExplicit(result, httpContext);";
+            codeWriter.WriteLine("if (result == null)");
+            codeWriter.StartBlock();
+            codeWriter.WriteLine("""throw new InvalidOperationException("The IResult returned by the Delegate must not be null.");""");
+            codeWriter.EndBlock();
+            codeWriter.WriteLine($"{returnOrAwait} GeneratedRouteBuilderExtensionsCore.ExecuteAsyncExplicit(result, httpContext);");
         }
         else if (endpointResponse.ResponseType?.SpecialType == SpecialType.System_String)
         {
-            return $"{returnOrAwait} httpContext.Response.WriteAsync(result);";
+            codeWriter.WriteLine($"{returnOrAwait} httpContext.Response.WriteAsync(result);");
         }
         else if (endpointResponse.ResponseType?.SpecialType == SpecialType.System_Object)
         {
-            return $"{returnOrAwait} GeneratedRouteBuilderExtensionsCore.ExecuteReturnAsync(result, httpContext, objectJsonTypeInfo);";
+            codeWriter.WriteLine($"{returnOrAwait} GeneratedRouteBuilderExtensionsCore.ExecuteReturnAsync(result, httpContext, objectJsonTypeInfo);");
         }
         else if (!endpointResponse.HasNoResponse)
         {
-            return $"{returnOrAwait} {endpointResponse.EmitJsonResponse()}";
+            codeWriter.WriteLine($"{returnOrAwait} {endpointResponse.EmitJsonResponse()}");
         }
         else if (!endpointResponse.IsAwaitable && endpointResponse.HasNoResponse)
         {
-            return $"{returnOrAwait} Task.CompletedTask;";
+            codeWriter.WriteLine($"{returnOrAwait} Task.CompletedTask;");
         }
         else
         {
-            return $"{returnOrAwait} httpContext.Response.WriteAsync(result);";
+            codeWriter.WriteLine($"{returnOrAwait} httpContext.Response.WriteAsync(result);");
         }
     }
 
@@ -366,7 +377,15 @@ internal static class StaticRouteHandlerModelEmitter
         }
         else if (endpoint.Response?.IsAwaitable == true)
         {
-            codeWriter.WriteLine($"var result = await handler({endpoint.EmitFilteredArgumentList()});");
+            codeWriter.WriteLine($"var task = handler({endpoint.EmitFilteredArgumentList()});");
+            if (endpoint.Response?.ResponseType?.NullableAnnotation == NullableAnnotation.Annotated)
+            {
+                codeWriter.WriteLine("if (task == null)");
+                codeWriter.StartBlock();
+                codeWriter.WriteLine("return (object?)Results.Empty;");
+                codeWriter.EndBlock();
+            }
+            codeWriter.WriteLine($"var result = await task;");
             codeWriter.WriteLine("return (object?)result;");
         }
         else

--- a/src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_ExplicitBodyParam_ComplexReturn_Snapshot.generated.txt
+++ b/src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_ExplicitBodyParam_ComplexReturn_Snapshot.generated.txt
@@ -120,6 +120,10 @@ namespace Microsoft.AspNetCore.Http.Generated
                         return;
                     }
                     var result = handler(todo_local!);
+                    if (result == null)
+                    {
+                        throw new InvalidOperationException("The IResult returned by the Delegate must not be null.");
+                    }
                     await GeneratedRouteBuilderExtensionsCore.ExecuteAsyncExplicit(result, httpContext);
                 }
 
@@ -220,6 +224,10 @@ namespace Microsoft.AspNetCore.Http.Generated
                         return;
                     }
                     var result = handler(todo_local);
+                    if (result == null)
+                    {
+                        throw new InvalidOperationException("The IResult returned by the Delegate must not be null.");
+                    }
                     await GeneratedRouteBuilderExtensionsCore.ExecuteAsyncExplicit(result, httpContext);
                 }
 

--- a/src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_ReturnsValidationProblemResult_Has_Metadata.generated.txt
+++ b/src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_ReturnsValidationProblemResult_Has_Metadata.generated.txt
@@ -109,6 +109,10 @@ namespace Microsoft.AspNetCore.Http.Generated
                         return Task.CompletedTask;
                     }
                     var result = handler();
+                    if (result == null)
+                    {
+                        throw new InvalidOperationException("The IResult returned by the Delegate must not be null.");
+                    }
                     return GeneratedRouteBuilderExtensionsCore.ExecuteAsyncExplicit(result, httpContext);
                 }
 

--- a/src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/Multiple_MapAction_NoParam_StringReturn.generated.txt
+++ b/src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/Multiple_MapAction_NoParam_StringReturn.generated.txt
@@ -184,7 +184,8 @@ namespace Microsoft.AspNetCore.Http.Generated
                         {
                             return (object?)Results.Empty;
                         }
-                        var result = await handler();
+                        var task = handler();
+                        var result = await task;
                         return (object?)result;
                     },
                     options.EndpointBuilder,
@@ -199,7 +200,8 @@ namespace Microsoft.AspNetCore.Http.Generated
                         httpContext.Response.StatusCode = 400;
                         return;
                     }
-                    var result = await handler();
+                    var task = handler();
+                    var result = await task;
                     if (result is string)
                     {
                         httpContext.Response.ContentType ??= "text/plain; charset=utf-8";
@@ -274,7 +276,8 @@ namespace Microsoft.AspNetCore.Http.Generated
                         {
                             return (object?)Results.Empty;
                         }
-                        var result = await handler();
+                        var task = handler();
+                        var result = await task;
                         return (object?)result;
                     },
                     options.EndpointBuilder,
@@ -289,7 +292,8 @@ namespace Microsoft.AspNetCore.Http.Generated
                         httpContext.Response.StatusCode = 400;
                         return;
                     }
-                    var result = await handler();
+                    var task = handler();
+                    var result = await task;
                     if (result is string)
                     {
                         httpContext.Response.ContentType ??= "text/plain; charset=utf-8";

--- a/src/Http/Http.Extensions/test/RequestDelegateGenerator/CompileTimeCreationTests.cs
+++ b/src/Http/Http.Extensions/test/RequestDelegateGenerator/CompileTimeCreationTests.cs
@@ -374,4 +374,35 @@ app.MapGet("/", () => "Hello world!");
         var exception = Assert.Throws<InvalidOperationException>(() => GetEndpointFromCompilation(compilation, serviceProvider: serviceProvider));
         Assert.Equal("JsonSerializerOptions instance must specify a TypeInfoResolver setting before being marked as read-only.", exception.Message);
     }
+
+    public static IEnumerable<object[]> NullResult
+    {
+        get
+        {
+            return new List<object[]>
+            {
+                new object[] { "IResult? () => null", "The IResult returned by the Delegate must not be null." },
+                new object[] { "Task<IResult?>? () => null", "The Task returned by the Delegate must not be null." },
+                new object[] { "Task<bool?>? () => null", "The Task returned by the Delegate must not be null." },
+                new object[] { "Task<IResult?> () => Task.FromResult<IResult?>(null)", "The IResult returned by the Delegate must not be null." },
+                new object[] { "ValueTask<IResult?> () => ValueTask.FromResult<IResult?>(null)", "The IResult returned by the Delegate must not be null." },
+            };
+        }
+    }
+
+    [Theory]
+    [MemberData(nameof(NullResult))]
+    public async Task RequestDelegateThrowsInvalidOperationExceptionOnNullDelegate(string innerSource, string message)
+    {
+        var source = $"""
+app.MapGet("/", {innerSource});
+""";
+        var (_, compilation) = await RunGeneratorAsync(source);
+        var endpoint = GetEndpointFromCompilation(compilation);
+
+        var httpContext = CreateHttpContext();
+        var exception = await Assert.ThrowsAsync<InvalidOperationException>(async () => await endpoint.RequestDelegate(httpContext));
+
+        Assert.Equal(message, exception.Message);
+    }
 }

--- a/src/Http/Http.Extensions/test/RequestDelegateGenerator/RequestDelegateCreationTests.Responses.cs
+++ b/src/Http/Http.Extensions/test/RequestDelegateGenerator/RequestDelegateCreationTests.Responses.cs
@@ -67,6 +67,9 @@ app.MapGet("/hello", () => "Hello world!")
     [InlineData(@"app.MapGet(""/"", () => 123456);", "123456")]
     [InlineData(@"app.MapGet(""/"", () => true);", "true")]
     [InlineData(@"app.MapGet(""/"", () => new DateTime(2023, 1, 1));", @"""2023-01-01T00:00:00""")]
+    [InlineData(@"app.MapGet(""/"", int () => 123456);", "123456")]
+    [InlineData(@"app.MapGet(""/"", bool () => true);", "true")]
+    [InlineData(@"app.MapGet(""/"", DateTime () => new DateTime(2023, 1, 1));", @"""2023-01-01T00:00:00""")]
     public async Task MapAction_NoParam_AnyReturn(string source, string expectedBody)
     {
         var (result, compilation) = await RunGeneratorAsync(source);
@@ -85,11 +88,17 @@ app.MapGet("/hello", () => "Hello world!")
     public static IEnumerable<object[]> MapAction_NoParam_ComplexReturn_Data => new List<object[]>()
     {
         new object[] { """app.MapGet("/", () => new Todo() { Name = "Test Item"});""" },
+        new object[] { """app.MapGet("/", Todo () => new Todo() { Name = "Test Item"});""" },
+        new object[] { """app.MapGet("/", Todo? () => new Todo() { Name = "Test Item"});""" },
         new object[] { """
 object GetTodo() => new Todo() { Name = "Test Item"};
 app.MapGet("/", GetTodo);
 """},
-        new object[] { """app.MapGet("/", () => TypedResults.Ok(new Todo() { Name = "Test Item"}));""" }
+        new object[] { """
+object? GetTodo() => new Todo() { Name = "Test Item"};
+app.MapGet("/", GetTodo);
+"""},
+        new object[] { """app.MapGet("/", IResult () => TypedResults.Ok(new Todo() { Name = "Test Item"}));""" }
     };
 
     [Theory]
@@ -138,7 +147,10 @@ app.MapGet("/", GetTodo);
     {
         new object[] { @"app.MapGet(""/"", () => Task.FromResult(""Hello world!""));", "Hello world!" },
         new object[] { @"app.MapGet(""/"", () => Task.FromResult(new Todo() { Name = ""Test Item"" }));", """{"id":0,"name":"Test Item","isComplete":false}""" },
-        new object[] { @"app.MapGet(""/"", () => Task.FromResult(TypedResults.Ok(new Todo() { Name = ""Test Item"" })));", """{"id":0,"name":"Test Item","isComplete":false}""" }
+        new object[] { @"app.MapGet(""/"", () => Task.FromResult(TypedResults.Ok(new Todo() { Name = ""Test Item"" })));", """{"id":0,"name":"Test Item","isComplete":false}""" },
+        new object[] { @"app.MapGet(""/"", Task<string> () => Task.FromResult(""Hello world!""));", "Hello world!" },
+        new object[] { @"app.MapGet(""/"", Task<Todo> () => Task.FromResult(new Todo() { Name = ""Test Item"" }));", """{"id":0,"name":"Test Item","isComplete":false}""" },
+        new object[] { @"app.MapGet(""/"", Task<Microsoft.AspNetCore.Http.HttpResults.Ok<Todo>> () => Task.FromResult(TypedResults.Ok(new Todo() { Name = ""Test Item"" })));", """{"id":0,"name":"Test Item","isComplete":false}""" }
     };
 
     [Theory]
@@ -890,10 +902,12 @@ app.MapPost("/", bool? () => null);
 
             var testTaskBoolAction = """
 app.MapPost("/", () => Task.FromResult<bool?>(null));
+app.MapPost("/", Task<bool?> () => Task.FromResult<bool?>(null));
 """;
 
             var testValueTaskBoolAction = """
 app.MapPost("/", () => ValueTask.FromResult<bool?>(null));
+app.MapPost("/", ValueTask<bool?> () => ValueTask.FromResult<bool?>(null));
 """;
 
             var testIntAction = """
@@ -902,22 +916,29 @@ app.MapPost("/", int? () => null);
 
             var testTaskIntAction = """
 app.MapPost("/", () => Task.FromResult<int?>(null));
+app.MapPost("/", Task<int?> () => Task.FromResult<int?>(null));
 """;
 
             var testValueTaskIntAction = """
 app.MapPost("/", () => ValueTask.FromResult<int?>(null));
+app.MapPost("/", ValueTask<int?> () => ValueTask.FromResult<int?>(null));
 """;
 
             var testTodoAction = """
+int id = 0;
+Todo? GetMaybeTodo() => id == 0 ? null : new Todo();
 app.MapPost("/", Todo? () => null);
+app.MapGet("/", GetMaybeTodo);
 """;
 
             var testTaskTodoAction = """
 app.MapPost("/", () => Task.FromResult<Todo?>(null));
+app.MapPost("/", Task<Todo?>? () => Task.FromResult<Todo?>(null));
 """;
 
             var testValueTaskTodoAction = """
 app.MapPost("/", () => ValueTask.FromResult<Todo?>(null));
+app.MapPost("/", ValueTask<Todo?> () => ValueTask.FromResult<Todo?>(null));
 """;
 
             var testTodoStructAction = """
@@ -945,12 +966,15 @@ app.MapPost("/", TodoStruct? () => null);
     public async Task RequestDelegateWritesNullReturnNullValue(string source)
     {
         var (_, compilation) = await RunGeneratorAsync(source);
-        var endpoint = GetEndpointFromCompilation(compilation);
+        var endpoints = GetEndpointsFromCompilation(compilation);
 
-        var httpContext = CreateHttpContext();
-        await endpoint.RequestDelegate(httpContext);
+        foreach (var endpoint in endpoints)
+        {
+            var httpContext = CreateHttpContext();
+            await endpoint.RequestDelegate(httpContext);
 
-        await VerifyResponseBodyAsync(httpContext, "null");
+            await VerifyResponseBodyAsync(httpContext, "null");
+        }
     }
 
     [Theory]

